src/libos/src/net/socket/unix/stream/socket.rs

```diff
@@ -0,0 +1,207 @@
+pub struct Stream {
+    addr: RwLock<Addr>,
+    end: SgxMutex<Option<Arc<EndPoint>>>, // Set after connection
+    is_blocking: AtomicBool,
+}
+
+impl Stream {
+    pub fn new(flags: FileFlags) -> Result<Self> {
+        Ok(Self {
+            addr: RwLock::new(Addr::Unnamed),
+            end: SgxMutex::new(None),
+            is_blocking: AtomicBool::new(!flags.contains(FileFlags::SOCK_NONBLOCK)),
+        })
+    }
+
+    pub fn socketpair(flags: FileFlags) -> Result<(Self, Self)> {
+        let is_blocking = !flags.contains(FileFlags::SOCK_NONBLOCK);
+        let (end_a, end_b) = EndPoint::end_pair()?;
+
+        let socket_a = Self {
+            addr: RwLock::new(Addr::Unnamed),
+            end: SgxMutex::new(Some(end_a)),
+            is_blocking: AtomicBool::new(is_blocking),
+        };
+
+        let socket_b = Self {
+            addr: RwLock::new(Addr::Unnamed),
+            end: SgxMutex::new(Some(end_b)),
+            is_blocking: AtomicBool::new(is_blocking),
+        };
+
+        Ok((socket_a, socket_b))
+    }
+
+    pub fn addr(&self) -> Addr {
+        self.addr.read().unwrap().clone()
+    }
+
+    pub fn end(&self) -> &SgxMutex<Option<Arc<EndPoint>>> {
+        &self.end
+    }
+
+    pub fn peer_addr(&self) -> Result<Addr> {
+        self.end
+            .lock()
+            .unwrap()
+            .as_ref()
+            .map(|end| end.peer_addr())
+            .flatten()
+            .ok_or_else(|| errno!(ENOTCONN, "the socket is not connected"))
+    }
+
+    /// Treat socket with its peer not dropped as connected, regardless of shutdown status.
+    pub fn is_connected(&self) -> bool {
+        if let Some(ref end) = *self.end.lock().unwrap() {
+            end.is_connected()
+        } else {
+            false
+        }
+    }
+
+    pub fn is_blocking(&self) -> bool {
+        self.is_blocking.load(Ordering::SeqCst)
+    }
+
+    pub fn set_non_blocking(&self) {
+        self.is_blocking.store(false, Ordering::SeqCst);
+        let end = self.end.lock().unwrap();
+        end.as_ref().map(|c| c.set_non_blocking());
+    }
+
+    pub fn set_blocking(&self) {
+        self.is_blocking.store(true, Ordering::SeqCst);
+        let end = self.end.lock().unwrap();
+        end.as_ref().map(|c| c.set_blocking());
+    }
+
+    // TODO: create the corresponding file in the fs
+    pub fn bind(&self, addr: &Addr) -> Result<()> {
+        if !self.addr().is_unnamed() {
+            return_errno!(EINVAL, "the socket is already bound");
+        }
+
+        SERVER_SPACE.add_binder(addr)?;
+
+        *self.addr.write().unwrap() = addr.clone();
+        if let Some(ref end) = *self.end.lock().unwrap() {
+            end.set_addr(&addr);
+        }
+        Ok(())
+    }
+
+    //TODO: add backlog support
+    pub fn listen(&self, backlog: i32) -> Result<()> {
+        if self.addr().is_unnamed() {
+            return_errno!(EINVAL, "the socket is not bound");
+        }
+
+        SERVER_SPACE.add_listener(&self.addr())
```

Comment:
What if the `listen` method is called on a connected socket? Without the aforementioned state machine, it is entirely unclear that a socket that is in a specific state behaves correctly.